### PR TITLE
[tk] Migrate tk to conan 2.0

### PR DESCRIPTION
**Pull request by samuel-emrys**
* Move to modern conan implementation
* Transition most file modifications to patches

Specify library name and version:  **tk/8.6.10**

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by samuel-emrys** [comment]
Relates to https://github.com/conan-io/conan-center-index/issues/17337

**Comment by samuel-emrys** [comment]
This is blocked pending merge of https://github.com/conan-io/conan-center-index/pull/17664 and associated update of `fontconfig` to use it

**Comment by ericLemanissier** [comment]
blocked by:
- https://github.com/conan-io/conan-center-index/pull/17664
- https://github.com/conan-io/conan-center-index/pull/17995

**Comment by ericLemanissier** [comment]
this one can be re-triggered now that https://github.com/conan-io/conan-center-index/pull/17995 is merged

**Comment by ericLemanissier** [comment]
blocked by:
- https://github.com/conan-io/conan-center-index/issues/18162

**Comment by samuel-emrys** [comment]
Before you spend too much time on this, #18569 is pretty much ready to merge, just waiting for broader libuuid issue to be resolved so that the 2.x tests pass. Appreciate the effort though! Probably worth scanning existing PRs before choosing another migration to save some duplicated effort

**Comment by samuel-emrys** [comment]
@jcar87 this is ready for review now

**Comment by mayeut** [comment]
@samuel-emrys, I added tk 8.6.13 on top of this in #19309, checking if macOS arm64 works with that version at the same time (I can build 8.6.10 locally on macOS arm64 by just discarding the last commit so I'm pessimistic that it'll work on cci CI with 8.6.13)

File: recipes/tk/all/conanfile.py
Comment (mayeut) [inline recipes/tk/all/conanfile.py:83]: It seems macOS m1 builds are cross-builds on CCI. A local build on M1 works just fine. I tried to cross-compile x86_64 from M1 and it fails.
```suggestion
        if self.settings.os == "Macos" and cross_building(self):
            raise ConanInvalidConfiguration("The tk conan recipe does not currently support Macos cross-builds. A contribution to add this functionality would be welcome.")
```

Comment (mayeut) [reply recipes/tk/all/conanfile.py:83]: cross-build fixed by https://github.com/conan-io/conan-center-index/pull/19309/commits/e7d3358a0a8dde9d80824050ecd59d0fee721e43

I'm not sure that's the correct way to do this but it works (at least locally, currently building on CCI CI).

Comment (samuel-emrys) [reply recipes/tk/all/conanfile.py:83]: @mayeut thanks for that. Happy to keep that out of scope in this PR and fix it on yours if the Conan team are too

**Review by AndreyMlashkin — APPROVED** [review]

**Review by EstebanDugueperoux2 — APPROVED** [review]

**Review by MartinDelille — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by RubenRBS — COMMENTED** [review]
Hi! Thanks for your contribution, patience, and for documenting the symbol issue that eventually led to libuuid's new recipe, really appreciated.

The hardest thing about this PR is to understand the implications of the patches and their effects on the build (beyond their description)

* I'd love to learn why adding CPPFLAGS makes this unix compatible, and more importantly, what's the reasoning for the comparison change in `unix/configure`

**Comment by samuel-emrys** [comment]
> I'd love to learn why adding CPPFLAGS makes this unix compatible, and more importantly, what's the reasoning for the comparison change in unix/configure

This is just an artifact of modernising the recipe for Conan 2. This modification to add CPPFLAGS already existed; this was just part of a simplification to move the existing inline patches out to a patch file to ease maintenance. See the removals in the [initial commit](https://github.com/conan-io/conan-center-index/commit/04c140ed5f285c1964f608273da4afae6543825d).

This is enigmatic of this entire PR - there should be no new functionality here, but it was a rather substantial change to make it broadly equivalent for Conan 2. I took the opportunity to simplify the recipe itself wherever I could.

**Comment by RubenRBS** [comment]
Oh I see! Sorry I only partially checked the PR, you're right in that the patches are already described in the old version, sorry to waste your time :)

After properly giving this a look, the rest of the PR I think looks good, will do some more testing locally before approving, but should get merged today :)

**Review by AndreyMlashkin — APPROVED** [review]

**Review by RubenRBS — APPROVED** [review]
Thanks a lot for your patience! :)

**Review by EstebanDugueperoux2 — APPROVED** [review]

**Pull request merged**
